Improvement/s3 c 9769/ignore trailing checksum

fredmnl opened the pull request:
# Motivation

Since aws-cli 2.23 (and associated SDK version), put-object now defaults to using unsigned payload with trailing checksums for integrity.

This is particularly nice from a computational point of view because the client only needs to read their file once to perform the upload. With AuthV4 signed payload, you would need to first compute the checksum of the object, to eventually include it in the signature contained in the header. With trailing checksums (and unsigned payload), the checksum is computed while the upload is happening and the checksum is sent at the end of the body. 

# Implementation

In this first iteration we are only interested in accepting the incoming requests using trailing checksums and unsigned payload. We are not yet verifying the checksum although the current proposal is extensible to perform this checksum verification as well.

fredmnl committed:
S3C-9769: Ignore trailing checksums in upload requests

bert-e commented:
# Hello fredmnl,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect Jira project



The Jira issue S3C-9769 specified in the source
branch name, does not belong to project `CLDSRV`.

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+                this.bytesToRead = 0;
+                continue;
+            }
+            const lineBreakIndex = chunk.indexOf('\r\n');
```
I think it needs to be:

```suggestion
const lineBreakIndex = this.chunkSizeBuffer.indexOf('\r\n');
```

Otherwise, it wouldn't be able to detect the newline with two chunks such as:

```
abcd\r  # chunk 1
\nefgh  # chunk 2
```

Then, you'll need to adapt the computed indices below with this change.

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+     */
+    _transform(chunkInput, encoding, callback) {
+        let chunk = chunkInput;
+        for (;;) {
```
go-ism? :wink: 

Maybe more idiomatic in JS: `while (true) {...}`

fredmnl replied:
Actually the linter does not like `while (true) {`, telling me that it does not like constant condition in a `while`. It does not pass the CI because it creates a warning. This `for (;;)` passes so :shrug:. What do you think?

jonathan-gramain replied:
you could then replace it with `while (chunk.byteLength !== 0) {...}`.

Otherwise, could use an `eslint-disable` rule if the `while (true)` is justified.

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+                    // chunk is smaller than the advertised size
+                    // forward the whole chunk
+                    this.bytesToRead -= chunk.byteLength;
+                    this.outputBuffer = Buffer.concat([this.outputBuffer, chunk]);
```
Performance-wise, it may be better to store an array of buffers, to store individual chunks as they are received, instead of concatenating buffers repeatedly, especially if chunking on application side is much larger than nodejs chunking (normally 64KB by default IIRC).

So initializing with:

```
this.outputBuffer = [];
```

Then pushing to it each chunk with

```
this.outputBuffer.push(chunk);
```

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+            this.bytesToRead = dataSize;
+        }
+
+        this.toFlush = this.outputBuffer;
```
Then to followup with `this.outputBuffer` being an array, instead you could repeatedly call `this.push()` on each item, so you don't have to concatenate buffers.

jonathan-gramain replied:
Actually you wouldn't even have to store the chunk references in `this.outputBuffer`, just `this.push()` them (or return them in callback) as they are received by `_transform`, until you reach the intended size.

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+        this.errCb = errCb;
+        this.chunkSizeBuffer = Buffer.alloc(0);
+        this.outputBuffer = Buffer.alloc(0);
+        this.toFlush = Buffer.alloc(0);
```
I don't think you need to store this field, as it just seems to be computed before being sent out in `_transform`.

fredmnl replied:
Good point!

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,92 @@
+            }
+            this.chunkSizeBuffer = Buffer.alloc(0);
+            if (dataSize === 0) {
+                // last chunk, no more data to read, the stream is closed
```
I would add a TODO to implement parsing and checking the trailing checksum here.

fredmnl commented:
@jonathan-gramain Thanks again for the review, I reworked the whole class to address your concerns. 

I thought that my tests were covering all of the tricky splits that could happen but actually the http request module used was squashing the chunks together before sending. I resorted to coding a unit tests for the class which is much neater. I might have overdone it but they run pretty quickly so we could keep them all. `2487 passing (389ms)`

I reworked the logic in a way which I think is much neater. I'm only detecting `\r` instead of \r\n, which can't be split across chunks. I'm not actually checking for the '\n', but I think that's acceptable as the number of bytes cannot contain `\r` anyway. (We have never checked the `\r\n` after data for example). I'm effectively looking for regex `[^\r]+\r..{number_of_bytes}..` until `0\r` is found. 

I think that eventually we will want to refactor V4Transform to integrate this as well as checksum checking.

jonathan-gramain commented on S3C-9776-repro.txt:
I suggest moving this file under `docs/` (possibly in a subdir of `docs`, but not sure how to name it), I feel the file is potentially useful but the top-level directory may not be the right place.

Or maybe somewhere in `tests/`?

fredmnl replied:
I'm sorry, I should have mentioned the DROP commits, I won't merge this file into the branch

jonathan-gramain commented on conf/lab-ca-csr.cnf:
I am not sure you intended to commit those files in `conf`?

fredmnl replied:
Same here, won't merge these files

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,98 @@
+        this.errCb = errCb;
+        this.chunkSizeBuffer = Buffer.alloc(0);
+        this.outputBuffer = [];
+        this.bytesToDiscard = 0; // when trailing /r/n are present, we discard them but they can be in different chunks
```
```suggestion
        this.bytesToDiscard = 0; // when trailing \r\n are present, we discard them but they can be in different chunks
```

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,98 @@
+            chunk = chunk.subarray(lineBreakIndex2);
+
+            // chunk-size is sent in hex
+            const dataSize = Number.parseInt(this.chunkSizeBuffer.toString(), 16);
```
Maybe not so important but as a refinement: we could validate that the string is indeed a well-formed hexadecimal integer until the end before parsing it into a number (e.g. with a regexp such as `/^[0-9a-fA-F]+$/.test(this.chunkSizeBuffer.toString())`), as `Number.parseInt` doesn't do full validation, in particular, it stops at the first invalid character encountered and returns the number parsed so far.

fredmnl replied:
Good point :+1:

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,98 @@
+        }
+
+        // TODO: push the stream into a checksum accumulator
+        const toFlush = Buffer.concat(this.outputBuffer);
```
To avoid concatenating the buffers, which I think requires copying memory around, I'd suggest to repeatedly call `this.push()` on each item of `this.outputBuffer` instead. We would then not return `toFlush` in the callback anymore.

This being said, we can even directly call `this.push()` in place of pushing to `this.outputBuffer` wherever it's done in this function, so we don't even have to keep that list.

fredmnl replied:
Ohhhh, I didn't understand when you wrote it above, I thought we needed to push the stream in the callback but `this.push` does the same thing and can be called multiple times. That's much neater, thanks!

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,98 @@
+                continue;
+            }
+
+            const lineBreakIndex2 = chunk.indexOf('\r');
```
It looks functional to me, I still have a slight preference for finding the full sequence `\r\n` in the `chunkSizeBuffer` because it also validates the stream contents more strongly. But it's not blocking for me, I'm fine with this approach too.

fredmnl replied:
I think I'd like to refactor both Transforms (the AuthV4 too) into a single processor that would actually work on a configurable FSM to both validate and process the stream. I think it would also make the V4Transform much more readable. With that in mind, we could deliver this iteration as is (we're aiming for a timely release on this fix). There will at some point be a further iteration to validate the checksums (prioritization aside)

jonathan-gramain commented on tests/unit/auth/TrailingChecksumTransform.js:
```diff
@@ -0,0 +1,109 @@
+        });
+    });
+
+    // test all trisection of the input string
```
While we're at it :smile: why not adding a single test where each chunk is a single byte?

fredmnl replied:
I don't know if you're kidding or not, but I love the idea, I'm using it :+1: 

Should we keep the trisection test? It's about 4000 cases but they run really fast, I don't know what the good practice is here.

jonathan-gramain replied:
I was not kidding :smile: 

Yeah I don't think the trisection tests hurt, especially if they all run in less than a second and they could catch some potential corner cases.

fredmnl commented:
@jonathan-gramain I think we're ready for another review here

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,100 @@
+            const lineBreakIndex = chunk.indexOf('\r');
+            if (lineBreakIndex === -1) {
+                if (this.chunkSizeBuffer.byteLength + chunk.byteLength > 10) {
+                    this.log.debug('chunk size field too big', {
```
Two suggestions:

Turn into `info` level, as it can be useful for debugging on production platforms in case their app sends chunks that are too big for us to support

You could do first:

```
this.chunkSizeBuffer = Buffer.concat([this.chunkSizeBuffer, chunk]);
```

then check that the size is below what we support.

Then you can log directly `this.chunkSizeBuffer.subarray(0, 16).toString()`:

- You can remove the hex encoding since it's already JSON-encoded in logs

- I think 16 chars allow to see the full size, while 8 chars may mask the least significant digits of the hex-encoded length

fredmnl replied:
(Also see comment below, which I wrote first)

`5GB` is an AWS limit, it's forbidden to perform a single upload of more than `5GB` hence the hard specification limit here, and `5GB` needs 9 hex characters to be written, we give it an extra character for `\r`. 

```
this.chunkSizeBuffer = Buffer.concat([this.chunkSizeBuffer, chunk]);
```

could potentially make us perform a huge buffer concatenation (although I think that Node would rechunk things before reaching this part of the code)

:+1: on hex

:+1: Yes 16 is better, in case the chunkSizeBuffer is empty, we would not see the whole 10 characters that we could have been interested in.

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,100 @@
+            }
+            const dataSize = Number.parseInt(this.chunkSizeBuffer.toString(), 16);
+            if (Number.isNaN(dataSize)) {
+                this.log.debug('unable to parse chunk size', {
```
This case now looks unlikely due to the sanity check, so I think you can log it with error level

fredmnl replied:
Sounds good, will do :+1:

jonathan-gramain commented on lib/auth/streamingV4/trailingChecksumTransform.js:
```diff
@@ -0,0 +1,100 @@
+                        chunkSizeBuffer: this.chunkSizeBuffer.toString('hex'),
+                        truncatedChunk: chunk.subarray(0, 8).toString('hex'),
+                    });
+                    // if bigger, the chunk would be over 5 GB
+                    // returning early to avoid a DoS by memory exhaustion
```
I believe this comment is outdated now that we don't store the entire chunk in memory, maybe simply state that it would be suspicious to have such a large chunk.

fredmnl replied:
That's not what's happening. If somehow a client is not sending a payload with trailing checksums (because it messed up the headers for some reason), we would be storing the payload in the `chunkSizeBuffer`. This could be done maliciously as well. 

This safeguard is what enables us to keep the `chunkSizeBuffer` in memory and `concat` to it, since we know it will be at most a few bytes.

jonathan-gramain commented on tests/unit/auth/TrailingChecksumTransform.js:
```diff
@@ -0,0 +1,131 @@
+            assert.strictEqual(err, null);
+        });
+        const chunks = [];
+        for (let i = 0; i < objDataWithTrailingChecksum.length - 1; i++) {
```
```suggestion
        for (let i = 0; i < objDataWithTrailingChecksum.length; i++) {
```

jonathan-gramain reviewed (approved):
LGTM with some minor suggestions

fredmnl commented:
/approve

bert-e commented:
# Incorrect Jira project



The Jira issue S3C-9769 specified in the source
branch name, does not belong to project `CLDSRV`.




*The following options are set:* **approve**

williamlardier reviewed (commented):
First review iteration, I will need to complete my review of the TrailingChecksumTransform class

fredmnl commented:
Closing this PR in favor of [https://github.com/scality/cloudserver/pull/5757](https://github.com/scality/cloudserver/pull/5757)
